### ISSUE TITLE
Index outputs dates as localized strings

### ISSUE THREAD

**Issue by hecon5** [joyfullservice/msaccess-vcs-addin#354]
While troubleshooting #351 performance issue, I discovered `clsVCSIndex.Update` and `.Save`, the import/export date is converted to a string before being stored in the dictionary `dItem` instead of letting `modJsonConverter.ParseJson` and `modJsonConverter.ConvertToJson` handle export and import of `dates`.

This will have some negative side effects, especially where the code base is stored on a traditional file server _AND_ developers reside in two different time zones (I've fallen prey to this one), or are simply using two different time format localizations, not to mention it may lead to inconsistencies if the `json` is handled outside of this tool (metric trackers, etc.).

This will not immediately affect instances using git where the index is stored only locally and not in the repository, single developer environments, or where all developers use the same time zone and localization. I would estimate the bulk of users fall in to the "this probably won't affect them" category. But for other users, this may have significant performance consequences if the index thinks the files were modified after export (leading/lagging time zones), conflict management false positives, etc.

In all cases below, for consistency, we should store dates as UTC dates to avoid any time zone issues, and convert to local only where needed, this ensures consistency across time zones and reduces conversions to/from local and UTC. Storing dates in ISO8601 format also ensures any exterior parser can determine if it's a particular time zone or UTC, and appropriately convert as needed to the desired localization.

Proposed fix options:
1. Store `dates` as `dates` and don't convert to strings. Caveat: need to check the performance impacts of storing a `date` vs `string`. Only convert during `ConvertToJson` exports, and let `modJsonConverter` handle it.
2. Store `dates` in ISO8601 compliant strings.
3. Use `modUtcConverter.ISO8601TimeStamp` to get timestamp strings. This includes millisecond accuracy and (from my testing) is fairly performant.
4.  `modUtcConverter.UTCMillisecondNow` (new function) to get a `date` to store, continue with use just like point 1 above. This would be (slightly) more performant than 3 in the cases of comparison, and would eventually (get converted to an ISO string anyway, but is (very slightly) slower than `VBA.Now()`, but has the advantage of already being converted to UTC, saving time during conversion to JSON.

**Comment by joyfullservice** [comment]
Thanks for sharing the detail behind this, and putting together the PR! I talked this over with one of my colleagues, and I think the benefits outweigh the cons from what I can see. I will go ahead and merge this in and continue with some performance testing on my end.

**Comment by hecon5** [comment]
No problem! One thing we've decided in the environments I support is that dates in use are stored always as local time values (example: `date` type is used is a "data in use") EXCEPT when explicitly defined as UTC (for instance, if you are finding the offset from UTC), AND the variable name clearly indicates it's UTC. 

Dates at rest (eg, once it's stored in a table on the back-end) the date is always stored as UTC date unless you use an ISO8601 compliant format (because then it's possible to determine offset to UTC or any arbitrary time zone). When that data is used on the front end, we convert to a local time for use, so the users don't know / notice / care, and don't get spooked by times of events at "midnight".

This way, there's no ambiguity, and programmers don't wonder what's what later on. Doing this has saved TONS of time debugging because our log times are always aligned, no matter who sends me the logs.

**Comment by joyfullservice** [comment]
Well, performance isn't looking real great at the moment... Parsing the ISO date becomes the largest bottleneck in scanning for changes, exceeding even the scanning of database objects. In some cases it nearly doubles the time required to scan for changes. Let me see if there is something we can optimize on the performance of this function...

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/ac936c0a-db81-42fd-b2f3-cf6f9c87ac6e)

**Comment by joyfullservice** [comment]
After digging into this, I think our biggest performance hit is that we are running two RegEx expressions on every date parsed. While RegEx is helpful in trying to cover a wide range of implementations for the ISO specification, in our case we really only need to parse a very specific known specification used in our application. These are not user-entered values, and will always be `YYYY-MM-DDTHH:NN:SS.###Z`. By streamlining the parsing to drop the RegEx and use more performant string functions, I think we could drastically improve the performance of reading these index records.

Let me see what I can do on that front...

**Comment by joyfullservice** [comment]
For reference, the Mac version of the `ParseIso` function doesn't use RegEx, and is 34 times faster!

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/c4513f98-6379-4145-bc66-78d8eb702c69)

However, this doesn't seem to be correctly parsing the date into the local time zone. It looks like this will need some further work.

**Comment by hecon5** [comment]
I don't have a Mac, so the Mac version is exactly from modUtcConverter (or pretty close anyway); I've since discovered there's a way to do it in mac, but there's also no Mac version of Access, so I wasn't too focused on building that out. See [this thread](https://www.mrexcel.com/board/threads/how-to-get-gmt-time-on-mac-with-vba.1117888/) for some information I didn't find before (because again, don't have a mac, there wasn't (isn't?) an Access version for mac, so it wasn't a hunt I sought).

It's interesting, the performance I get from `ParseJson` is considerably faster than what you're showing, why is not quite clear to me...I also do not see any speed hits (or at least nowhere near as painful as you're showing)... it's also the same speed as the prior method; I'm wondering if it's the same symptom you noted in #360 when I tried to update the JsonConverter. I'm going to do some testing on my end and determine what's going on here.

Another thought: we are only talking about seconds here with several thousand iterations...are we chasing something for the sake of chasing? 

> By streamlining the parsing to drop the RegEx and use more performant string functions, I think we could drastically improve the performance of reading these index records.

I agree; if we do we should note the customization of the tool since the functions as-is are mostly inline with vba-tools upstream with a few functionality improvements. I've got an idea on that; I'll make a new PR and branch so we can collaborate on it if you'd like.

**Comment by hecon5** [comment]
Some thoughts out loud: 
- Parse the date part with regular string tools, since we're only going to see the input from json, this seems straightforward enough.
- Parse the time part with regular string tools. Note we'll need to validate the time zone offset is not present. 
- I agree with your thoughts on getting rid of the regex if possible; I needed a way in our ap to tolerate multiple date inputs and since we don't call it 1000's of times in a row, the utility of parsing any date type was considerably more important than speed. 
- I think if we're going to create a custom implementation, we should use `clsConcat` as it will simplify some items and ensure consistent string concatenation tooling within the whole Addin.

**Comment by joyfullservice** [comment]
Okay, it looks like it works as well as I was hoping. Essentially I just added a couple wrapper functions that attempt a high-performance conversion first, then fall back to the more comprehensive RegEx approach if necessary. Now the date parsing time is back down to about 0.07 seconds over 3,300 iterations, which is acceptable to me.

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/7fc323aa-e241-4c9c-b5e9-eee0145539c6)

**Comment by hecon5** [comment]
Sounds good; I see you've added a hard-coded "-" for  split. I'd suggest using `ISO8601DateDelimiter` since it's already in there and ensures a standard definition in case (shudder) the standard changes. 

```diff
--- 722    varParts = Split(InVal, "-")
+++ 722    varParts = Split(InVal, ISO8601DateDelimiter)
```

**Comment by hecon5** [comment]
What's interesting is the date was the outlier; you didn't do anything about the time; so that's a curiosity to me for why the date but not the time causes that much performance hit.

**Comment by joyfullservice** [comment]
> Sounds good; I see you've added a hard-coded "-" for split. I'd suggest using `ISO8601DateDelimiter` since it's already in there and ensures a standard definition in case (shudder) the standard changes.
> 
> ```diff
> --- 722    varParts = Split(InVal, "-")
> +++ 722    varParts = Split(InVal, ISO8601DateDelimiter)
> ```

For me, I think it comes down to simplicity and style. In the following code sample, I personally find the first approach more intuitively obvious to read and understand. In the second option, I have to stop and think about what ISO8601 means, what that date delimiter might look like, and may need to check the constant value to see if it matches the date I am parsing. Add to that the low likelihood that this value will change in the global standard, and I lean towards simplicity for something like this. As Mike Wolf often [reminds](https://nolongerset.com/avoid-context-switching/) us, it is easier to maintain code that is easy to understand.

```VBA
    If InVal Like "####-##-##" Then
        ' Use high-performance conversion to date
        varParts = Split(InVal, "-")
    End If
    
    If InVal Like "####" & ISO8601DateDelimiter & "##" & ISO8601DateDelimiter & "##" Then
        ' Use high-performance conversion to date
        varParts = Split(InVal, ISO8601DateDelimiter)
    End If
```

Again, there is nothing wrong with using constants. Both approaches are good and work perfectly fine. I just wrote the code, so I got to pick.  :grinning:

**Comment by hecon5** [comment]
Closed by #458
